File: src/validate.ts
// Manifest linter — `kcp-agent validate` for knowledge.yaml publishers.
//
// Validates the same compact model the planner consumes, so "validates clean"
// means "this agent can navigate it". Errors are structural problems that will
// mislead or fail an agent; warnings are declarations that weaken navigation
// (empty triggers, stale temporal blocks) but don't break it.

import { existsSync } from "node:fs";
import { dirname, join, isAbsolute } from "node:path";
import { loadManifestText, parseManifest } from "./client.js";
import type { FetchGuard } from "./fetch.js";
import { terms } from "./planner.js";
import type { Manifest, Unit } from "./model.js";

export interface Finding {
  level: "error" | "warning";
  where: string; // e.g. "unit 'deploy-guide'" or "manifest"
  message: string;
}

export interface ValidationReport {
  source: string;
  project?: string;
  findings: Finding[];
  /** No errors (warnings allowed). */
  ok: boolean;
}

const ACCESS_VALUES = new Set(["public", "authenticated", "restricted"]);
/**
 * The only `kind` the planner treats as governed (spec §4.3a, v0.26.1). Any
 * other non-empty value — a typo ("Skill", "skil") or an unsupported class
 * ("procedure") — silently falls through the planner's `unit.kind === "skill"`
 * check as an ordinary knowledge unit: the skill_eligibility gate never
 * applies, and action_scope/load_eligible are ignored. That's an
 * unknown-data-passes-through gap in an otherwise fail-closed gate, so flag
 * it at publish time instead of letting it load silently ungoverned.
 */
const KIND_VALUES = new Set(["skill"]);
const ISO_DATE = /^\d{4}-\d{2}-\d{2}([T ].*)?$/;

function unsafePath(path: string): string | undefined {
  if (path === "") return "path is empty";
  if (/^[a-z][a-z0-9+.-]*:/i.test(path) || path.startsWith("//")) return "path must be relative, not a URL";
  if (isAbsolute(path)) return "path must be relative, not absolute";
  if (path.split("/").includes("..")) return "path must not traverse with '..'";
  return undefined;
}

/** Validate a parsed manifest. `baseDir` enables unit-path existence checks. */
export function validateManifest(manifest: Manifest, baseDir?: string): Finding[] {
  const findings: Finding[] = [];
  const err = (where: string, message: string) => findings.push({ level: "error", where, message });
  const warn = (where: string, message: string) => findings.push({ level: "warning", where, message });

  if (manifest.project === "(unnamed)") warn("manifest", "missing 'project'");
  if (manifest.version === "0.0.0") warn("manifest", "missing 'version'");
  if (!manifest.kcp_version) warn("manifest", "missing 'kcp_version' — agents cannot tell which spec revision this targets");
  if (manifest.units.length === 0) warn("manifest", "declares no units — nothing for an agent to navigate");

  const ids = new Set<string>();
  for (const unit of manifest.units) {
    const where = `unit '${unit.id || "(no id)"}'`;
    if (!unit.id) err(where, "missing 'id'");
    else if (ids.has(unit.id)) err(where, "duplicate unit id");
    ids.add(unit.id);

    const pathProblem = unsafePath(unit.path);
    if (pathProblem) err(where, pathProblem);
    else if (baseDir !== undefined && !existsSync(join(baseDir, unit.path))) {
      err(where, `path '${unit.path}' does not exist`);
    }

    if (!unit.intent) err(where, "missing 'intent' — intent is the primary navigation signal");
    if (unit.triggers.length === 0) warn(where, "no 'triggers' — unit is only findable through its intent text");
    if (unit.audience.length === 0) warn(where, "no 'audience' — declare who this unit serves (e.g. [agent, human])");
    if (unit.access && !ACCESS_VALUES.has(unit.access)) {
      warn(where, `unknown access '${unit.access}' (expected public/authenticated/restricted)`);
    }
    if (unit.kind && !KIND_VALUES.has(unit.kind)) {
      warn(where, `unknown kind '${unit.kind}' (expected 'skill') — will load as an ungoverned ` +
        `plain knowledge unit: action_scope/load_eligible/skill_eligibility will NOT apply to it`);
    }
    validateTemporal(unit, where, findings);
    validateNotFor(unit, where, findings);
    validateSkillScope(unit, where, findings);
    for (const m of unit.payment?.methods ?? []) {
      if (!m.type) err(where, "payment method missing 'type'");
      if (m.type === "x402" && (!m.price_per_request || !m.currency)) {
        warn(where, "x402 payment method should declare 'price_per_request' and 'currency'");
      }
    }
  }
  // superseded_by references are checked after all ids are collected
  for (const unit of manifest.units) {
    const succ = unit.temporal?.superseded_by;
    if (succ && !ids.has(succ)) {
      err(`unit '${unit.id}'`, `temporal.superseded_by references unknown unit '${succ}'`);
    }
  }

  const refIds = new Set<string>();
  for (const ref of manifest.manifests) {
    const where = `manifest ref '${ref.id || "(no id)"}'`;
    if (!ref.id) err(where, "missing 'id'");
    else if (refIds.has(ref.id)) err(where, "duplicate manifest ref id");
    refIds.add(ref.id);
    if (!ref.url) err(where, "missing 'url'");
    else if (!/^https:\/\//.test(ref.url)) warn(where, `url is not https — agents should fetch federation over TLS`);
    if (ref.agent_identity?.required && !ref.agent_identity.credential_hint) {
      warn(where, "agent_identity.required without 'credential_hint' — agents cannot plan credential acquisition");
    }
  }

  const ar = manifest.trust?.agent_requirements;
  if (ar?.require_attestation && (ar.trusted_providers ?? []).length === 0) {
    err("manifest", "require_attestation with no trusted_providers — no agent can ever qualify (permanently fail-closed)");
  }

  if (manifest.signing && manifest.signing.scheme && !/^(ed25519|eddsa)$/i.test(manifest.signing.scheme)) {
    warn("manifest", `signing scheme '${manifest.signing.scheme}' is not one this agent can verify (ed25519)`);
  }

  // Serving Endpoint Binding (§3.12): every entry MUST be an absolute https://
  // URL — an http:// entry is a §7 validation error, not a warning.
  if (manifest.serving) {
    const lists: [string, string[] | undefined][] = [
      ["serving.manifest", manifest.serving.manifest],
      ["serving.mcp", manifest.serving.mcp],
    ];
    for (const [field, list] of lists) {
      for (const entry of list ?? []) {
        if (!/^https:\/\//.test(entry)) {
          err("manifest", `${field} entry '${entry}' must be an absolute https:// URL (§3.12)`);
        } else {
          try {
            new URL(entry);
          } catch {
            err("manifest", `${field} entry '${entry}' is not a valid URL (§3.12)`);
          }
        }
      }
    }
    if (!manifest.signing) {
      warn("manifest", "serving block declared without a signing block — an unsigned binding can be stripped or altered by a re-hoster (§3.12 recommends signing)");
    }
  }

  return findings;
}

/**
 * The self-sabotaging gate: at plan time a `not_for` entry gates the unit
 * whenever any task term appears inside it. The most natural questions for a
 * unit are phrased in the unit's own vocabulary — so a `not_for` written as a
 * natural-language negation ("questions about non-AI software systems") that
 * contains the unit's own intent/trigger terms deterministically locks the
 * gate on exactly the audience the unit exists to serve. Found live in a
 * production regulatory manifest; a machine can flag it at publish time.
 */
function validateNotFor(unit: Unit, where: string, findings: Finding[]) {
  const notFor = unit.not_for ?? [];
  if (notFor.length === 0) return;
  const vocabulary = new Set<string>([...terms(unit.intent), ...unit.triggers.flatMap((t) => terms(t))]);
  for (const nf of notFor) {
    const entry = nf.toLowerCase();
    const hits = [...vocabulary].filter((v) => entry.includes(v)).sort();
    if (hits.length > 0) {
      findings.push({
        level: "warning",
        where,
        message:
          `not_for '${nf}' contains the unit's own vocabulary (${hits.join(", ")}) — ` +
          `term matching will gate this unit against its most natural questions; ` +
          `name the excluded topic in its own words (e.g. "CCPA", "accounting"), never as a negation of this unit's topic ("non-X", "outside X")`,
      });
    }
  }
}

/**
 * Governed-skill conformance (#100): a `kind: skill` unit is an invoke-eligible
 * procedure whose `action_scope` is the *only* thing that authorizes it to touch
 * a tool or path. Skills fail closed — so a skill that is granted invoke
 * eligibility (`load_eligible: true`) but declares no action_scope (or an empty
 * one) authorizes *nothing*: a publisher error that hands an autonomous agent a
 * runnable procedure with no declared blast radius. Flag it at publish time as
 * an error, the same fail-closed posture the planner takes at plan time.
 */
function validateSkillScope(unit: Unit, where: string, findings: Finding[]) {
  if (unit.kind !== "skill") return;
  const as = unit.action_scope;
  const empty =
    !as ||
    ((as.tools ?? []).length === 0 &&
      (as.paths ?? []).length === 0 &&
      (as.capabilities ?? []).length === 0);
  if (unit.load_eligible === true && empty) {
    findings.push({
      level: "error",
      where,
      message:
        "kind: skill is load_eligible but declares no action_scope — a governed skill authorizes nothing " +
        "until it lists the tools/paths/capabilities it may touch (fail-closed, #100)",
    });
  } else if (empty) {
    findings.push({
      level: "warning",
      where,
      message:
        "kind: skill declares no action_scope — it authorizes nothing and can never be invoked until it " +
        "declares one (fail-closed, #100)",
    });
  }
}

function validateTemporal(unit: Unit, where: string, findings: Finding[]) {
  const t = unit.temporal;
  if (!t) return;
  const err = (message: string) => findings.push({ level: "error", where, message });
  const warn = (message: string) => findings.push({ level: "warning", where, message });
  if (t.valid_from && !ISO_DATE.test(t.valid_from)) err(`temporal.valid_from '${t.valid_from}' is not an ISO date`);
  if (t.valid_until && !ISO_DATE.test(t.valid_until)) err(`temporal.valid_until '${t.valid_until}' is not an ISO date`);
  if (t.valid_from && t.valid_until && t.valid_until < t.valid_from) {
    err(`temporal window ends (${t.valid_until}) before it starts (${t.valid_from})`);
  }
  const today = new Date().toISOString().slice(0, 10);
  if (t.valid_until && ISO_DATE.test(t.valid_until) && t.valid_until < today && !t.superseded_by) {
    warn(`expired ${t.valid_until} with no 'superseded_by' — agents get a dead end instead of a successor`);
  }
}

/** Load a manifest from a path/dir/URL and validate it. */
export async function validateLocation(location: string, fetchGuard: FetchGuard = {}): Promise<ValidationReport> {
  let text: string;
  let source: string;
  try {
    ({ text, source } = await loadManifestText(location, fetchGuard));
  } catch (e) {
    const message = e instanceof Error ? e.message : String(e);
    return { source: location, findings: [{ level: "error", where: "manifest", message }], ok: false };
  }
  let manifest: Manifest;
  try {
    manifest = parseManifest(text, source);
  } catch (e) {
    const message = e instanceof Error ? e.message : String(e);
    return { source, findings: [{ level: "error", where: "manifest", message: `does not parse: ${message}` }], ok: false };
  }
  const baseDir = /^https?:\/\//.test(source) ? undefined : dirname(source);
  const findings = validateManifest(manifest, baseDir);
  return {
    source,
    project: manifest.project,
    findings,
    ok: !findings.some((f) => f.level === "error"),
  };
}

File: rust/kcp-planner/src/validate.rs
//! Manifest linter — a Rust port of `src/validate.ts`. Validates the same compact
//! model the planner consumes: errors are structural problems that mislead or
//! fail an agent; warnings are declarations that weaken navigation. Date/URL
//! checks are hand-rolled to keep the dependency tree minimal (no regex crate).

use crate::model::{parse_manifest, Manifest, Unit};
use crate::planner::terms;
use serde::Serialize;
use std::collections::HashSet;

#[derive(Debug, Clone, PartialEq)]
pub struct Finding {
    pub level: String, // "error" | "warning"
    pub where_: String,
    pub message: String,
}

// `where` is a Rust keyword; rename on the wire to match the TS field.
impl Finding {
    fn error(where_: &str, message: impl Into<String>) -> Finding {
        Finding { level: "error".to_string(), where_: where_.to_string(), message: message.into() }
    }
    fn warning(where_: &str, message: impl Into<String>) -> Finding {
        Finding { level: "warning".to_string(), where_: where_.to_string(), message: message.into() }
    }
}

// Custom serde to emit/read `where` (keyword) with the wire name.
mod finding_serde {
    use super::Finding;
    use serde::ser::SerializeStruct;
    use serde::{Deserialize, Deserializer, Serialize, Serializer};

    impl Serialize for Finding {
        fn serialize<S: Serializer>(&self, s: S) -> Result<S::Ok, S::Error> {
            let mut st = s.serialize_struct("Finding", 3)?;
            st.serialize_field("level", &self.level)?;
            st.serialize_field("where", &self.where_)?;
            st.serialize_field("message", &self.message)?;
            st.end()
        }
    }

    #[derive(Deserialize)]
    struct FindingWire {
        level: String,
        #[serde(rename = "where")]
        where_: String,
        message: String,
    }

    impl<'de> Deserialize<'de> for Finding {
        fn deserialize<D: Deserializer<'de>>(d: D) -> Result<Finding, D::Error> {
            let w = FindingWire::deserialize(d)?;
            Ok(Finding { level: w.level, where_: w.where_, message: w.message })
        }
    }
}

#[derive(Debug, Clone, Serialize)]
pub struct ValidationReport {
    pub source: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub project: Option<String>,
    pub findings: Vec<Finding>,
    pub ok: bool,
}

const ACCESS_VALUES: [&str; 3] = ["public", "authenticated", "restricted"];

/// Matches the TS `^[a-z][a-z0-9+.-]*:`i scheme test and `//` prefix.
fn looks_like_url(path: &str) -> bool {
    if path.starts_with("//") {
        return true;
    }
    let bytes = path.as_bytes();
    if bytes.is_empty() || !bytes[0].is_ascii_alphabetic() {
        return false;
    }
    let mut i = 1;
    while i < bytes.len() {
        let c = bytes[i];
        if c == b':' {
            return true;
        }
        if !(c.is_ascii_alphanumeric() || c == b'+' || c == b'.' || c == b'-') {
            return false;
        }
        i += 1;
    }
    false
}

fn unsafe_path(path: &str) -> Option<String> {
    if path.is_empty() {
        return Some("path is empty".to_string());
    }
    if looks_like_url(path) {
        return Some("path must be relative, not a URL".to_string());
    }
    if std::path::Path::new(path).is_absolute() || path.starts_with('/') {
        return Some("path must be relative, not absolute".to_string());
    }
    if path.split('/').any(|seg| seg == "..") {
        return Some("path must not traverse with '..'".to_string());
    }
    None
}

/// Today (UTC) as YYYY-MM-DD — epoch seconds → civil date (Howard Hinnant). Used
/// by callers that lint without an injected clock (the MCP server).
pub fn today_utc() -> String {
    let secs = std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).map(|d| d.as_secs() as i64).unwrap_or(0);
    let z = secs.div_euclid(86400) + 719468;
    let era = if z >= 0 { z } else { z - 146096 } / 146097;
    let doe = z - era * 146097;
    let yoe = (doe - doe / 1460 + doe / 36524 - doe / 146096) / 365;
    let y = yoe + era * 400;
    let doy = doe - (365 * yoe + yoe / 4 - yoe / 100);
    let mp = (5 * doy + 2) / 153;
    let d = doy - (153 * mp + 2) / 5 + 1;
    let m = if mp < 10 { mp + 3 } else { mp - 9 };
    let year = y + if m <= 2 { 1 } else { 0 };
    format!("{:04}-{:02}-{:02}", year, m, d)
}

/// Matches `^\d{4}-\d{2}-\d{2}([T ].*)?$`.
fn is_iso_date(s: &str) -> bool {
    let b = s.as_bytes();
    if b.len() < 10 {
        return false;
    }
    let digit = |i: usize| b[i].is_ascii_digit();
    if !(digit(0) && digit(1) && digit(2) && digit(3) && b[4] == b'-' && digit(5) && digit(6) && b[7] == b'-' && digit(8) && digit(9)) {
        return false;
    }
    b.len() == 10 || b[10] == b'T' || b[10] == b' '
}

/// Validate a parsed manifest. `base_dir` enables unit-path existence checks.
pub fn validate_manifest(manifest: &Manifest, base_dir: Option<&str>, today: &str) -> Vec<Finding> {
    let mut f: Vec<Finding> = Vec::new();

    if manifest.project == "(unnamed)" {
        f.push(Finding::warning("manifest", "missing 'project'"));
    }
    if manifest.version == "0.0.0" {
        f.push(Finding::warning("manifest", "missing 'version'"));
    }
    if manifest.kcp_version.is_none() {
        f.push(Finding::warning("manifest", "missing 'kcp_version' — agents cannot tell which spec revision this targets"));
    }
    if manifest.units.is_empty() {
        f.push(Finding::warning("manifest", "declares no units — nothing for an agent to navigate"));
    }

    let mut ids: HashSet<String> = HashSet::new();
    for unit in &manifest.units {
        let where_ = format!("unit '{}'", if unit.id.is_empty() { "(no id)" } else { &unit.id });
        if unit.id.is_empty() {
            f.push(Finding::error(&where_, "missing 'id'"));
        } else if ids.contains(&unit.id) {
            f.push(Finding::error(&where_, "duplicate unit id"));
        }
        ids.insert(unit.id.clone());

        if let Some(problem) = unsafe_path(&unit.path) {
            f.push(Finding::error(&where_, problem));
        } else if let Some(dir) = base_dir {
            if !std::path::Path::new(dir).join(&unit.path).exists() {
                f.push(Finding::error(&where_, format!("path '{}' does not exist", unit.path)));
            }
        }

        if unit.intent.is_empty() {
            f.push(Finding::error(&where_, "missing 'intent' — intent is the primary navigation signal"));
        }
        if unit.triggers.is_empty() {
            f.push(Finding::warning(&where_, "no 'triggers' — unit is only findable through its intent text"));
        }
        if unit.audience.is_empty() {
            f.push(Finding::warning(&where_, "no 'audience' — declare who this unit serves (e.g. [agent, human])"));
        }
        if let Some(access) = &unit.access {
            if !ACCESS_VALUES.contains(&access.as_str()) {
                f.push(Finding::warning(&where_, format!("unknown access '{}' (expected public/authenticated/restricted)", access)));
            }
        }
        validate_temporal(unit, &where_, today, &mut f);
        validate_not_for(unit, &where_, &mut f);
        for m in unit.payment.as_ref().and_then(|p| p.methods.as_ref()).map(|v| v.as_slice()).unwrap_or(&[]) {
            if m.r#type.is_empty() {
                f.push(Finding::error(&where_, "payment method missing 'type'"));
            }
            if m.r#type == "x402" && (m.price_per_request.is_none() || m.currency.is_none()) {
                f.push(Finding::warning(&where_, "x402 payment method should declare 'price_per_request' and 'currency'"));
            }
        }
    }

    for unit in &manifest.units {
        if let Some(succ) = unit.temporal.as_ref().and_then(|t| t.superseded_by.as_ref()) {
            if !ids.contains(succ) {
                f.push(Finding::error(&format!("unit '{}'", unit.id), format!("temporal.superseded_by references unknown unit '{}'", succ)));
            }
        }
    }

    let mut ref_ids: HashSet<String> = HashSet::new();
    for r in &manifest.manifests {
        let where_ = format!("manifest ref '{}'", if r.id.is_empty() { "(no id)" } else { &r.id });
        if r.id.is_empty() {
            f.push(Finding::error(&where_, "missing 'id'"));
        } else if ref_ids.contains(&r.id) {
            f.push(Finding::error(&where_, "duplicate manifest ref id"));
        }
        ref_ids.insert(r.id.clone());
        if r.url.is_empty() {
            f.push(Finding::error(&where_, "missing 'url'"));
        } else if !r.url.starts_with("https://") {
            f.push(Finding::warning(&where_, "url is not https — agents should fetch federation over TLS"));
        }
        if let Some(ai) = &r.agent_identity {
            if ai.required == Some(true) && ai.credential_hint.is_none() {
                f.push(Finding::warning(&where_, "agent_identity.required without 'credential_hint' — agents cannot plan credential acquisition"));
            }
        }
    }

    let ar = manifest.trust.as_ref().and_then(|t| t.agent_requirements.as_ref());
    if let Some(ar) = ar {
        if ar.require_attestation == Some(true) && ar.trusted_providers.is_empty() {
            f.push(Finding::error("manifest", "require_attestation with no trusted_providers — no agent can ever qualify (permanently fail-closed)"));
        }
    }

    if let Some(sig) = &manifest.signing {
        if let Some(scheme) = &sig.scheme {
            let low = scheme.to_lowercase();
            if low != "ed25519" && low != "eddsa" {
                f.push(Finding::warning("manifest", format!("signing scheme '{}' is not one this agent can verify (ed25519)", scheme)));
            }
        }
    }

    f
}

fn validate_not_for(unit: &Unit, where_: &str, f: &mut Vec<Finding>) {
    if unit.not_for.is_empty() {
        return;
    }
    let mut vocab: HashSet<String> = HashSet::new();
    for t in terms(&unit.intent) {
        vocab.insert(t);
    }
    for trig in &unit.triggers {
        for t in terms(trig) {
            vocab.insert(t);
        }
    }
    for nf in &unit.not_for {
        let entry = nf.to_lowercase();
        let mut hits: Vec<String> = vocab.iter().filter(|v| entry.contains(v.as_str())).cloned().collect();
        hits.sort();
        if !hits.is_empty() {
            f.push(Finding::warning(
                where_,
                format!(
                    "not_for '{}' contains the unit's own vocabulary ({}) — term matching will gate this unit against its most natural questions; name the excluded topic in its own words (e.g. \"CCPA\", \"accounting\"), never as a negation of this unit's topic (\"non-X\", \"outside X\")",
                    nf,
                    hits.join(", ")
                ),
            ));
        }
    }
}

fn validate_temporal(unit: &Unit, where_: &str, today: &str, f: &mut Vec<Finding>) {
    let t = match &unit.temporal {
        None => return,
        Some(t) => t,
    };
    if let Some(vf) = &t.valid_from {
        if !is_iso_date(vf) {
            f.push(Finding::error(where_, format!("temporal.valid_from '{}' is not an ISO date", vf)));
        }
    }
    if let Some(vu) = &t.valid_until {
        if !is_iso_date(vu) {
            f.push(Finding::error(where_, format!("temporal.valid_until '{}' is not an ISO date", vu)));
        }
    }
    if let (Some(vf), Some(vu)) = (&t.valid_from, &t.valid_until) {
        if vu < vf {
            f.push(Finding::error(where_, format!("temporal window ends ({}) before it starts ({})", vu, vf)));
        }
    }
    if let Some(vu) = &t.valid_until {
        if is_iso_date(vu) && vu.as_str() < today && t.superseded_by.is_none() {
            f.push(Finding::warning(where_, format!("expired {} with no 'superseded_by' — agents get a dead end instead of a successor", vu)));
        }
    }
}

/// Load a manifest from a local path/dir and validate it. `today` is injected
/// (the one clock the linter reads, for the expired-without-successor warning).
pub fn validate_location(location: &str, today: &str) -> ValidationReport {
    let (text, source) = match load_local_manifest_text(location) {
        Ok(v) => v,
        Err(e) => return ValidationReport { source: location.to_string(), project: None, findings: vec![Finding::error("manifest", e)], ok: false },
    };
    let manifest = match parse_manifest(&text, Some(&source)) {
        Ok(m) => m,
        Err(e) => return ValidationReport { source: source.clone(), project: None, findings: vec![Finding::error("manifest", format!("does not parse: {}", e))], ok: false },
    };
    let base_dir = std::path::Path::new(&source).parent().map(|p| p.to_string_lossy().to_string());
    let findings = validate_manifest(&manifest, base_dir.as_deref(), today);
    let ok = !findings.iter().any(|f| f.level == "error");
    ValidationReport { source, project: Some(manifest.project.clone()), findings, ok }
}

/// Resolve a path or directory (knowledge.yaml / .well-known/knowledge.yaml) to text + source.
/// The `source` label must match the TS reference (`loadManifestText` in client.ts),
/// which builds it with Node's `path.join` — so we normalize the same way (e.g.
/// `join(".", "knowledge.yaml")` yields `knowledge.yaml`, not `./knowledge.yaml`).
pub fn load_local_manifest_text(location: &str) -> Result<(String, String), String> {
    let source = if std::path::Path::new(location).is_dir() {
        let a = node_join(&[location, "knowledge.yaml"]);
        let b = node_join(&[location, ".well-known", "knowledge.yaml"]);
        if std::path::Path::new(&a).exists() {
            a
        } else if std::path::Path::new(&b).exists() {
            b
        } else {
            return Err(format!("no knowledge.yaml found in {}", location));
        }
    } else {
        location.to_string()
    };
    if !std::path::Path::new(&source).exists() {
        return Err(format!("manifest not found: {}", source));
    }
    let text = std::fs::read_to_string(&source).map_err(|e| e.to_string())?;
    Ok((text, source))
}

/// Join path segments the way Node's `path.join` does on POSIX: concatenate with
/// `/`, then normalize (drop `.` segments, resolve `..`). The TS reference records
/// the joined result as the manifest `source`, so byte-parity of `validate --json`
/// depends on matching this — `join(".", "knowledge.yaml")` → `knowledge.yaml`.
fn node_join(parts: &[&str]) -> String {
    let joined = parts.iter().filter(|p| !p.is_empty()).copied().collect::<Vec<_>>().join("/");
    node_normalize(&joined)
}

fn node_normalize(path: &str) -> String {
    let is_abs = path.starts_with('/');
    let mut stack: Vec<&str> = Vec::new();
    for seg in path.split('/') {
        match seg {
            "" | "." => {}
            ".." => match stack.last() {
                Some(&last) if last != ".." => {
                    stack.pop();
                }
                None if is_abs => {}
                _ => stack.push(".."),
            },
            s => stack.push(s),
        }
    }
    let body = stack.join("/");
    if is_abs {
        format!("/{}", body)
    } else if body.is_empty() {
        ".".to_string()
    } else {
        body
    }
}
